pina/loss.py

```diff
@@ -0,0 +1,182 @@
+        :type loss: torch.Tensor
+        :return: Reduced loss.
+        :rtype: torch.Tensor
+        """
+        if self.reduction == "none":
+            ret = loss
+        elif self.reduction == "mean":
+            ret = torch.mean(loss, keepdim=True, dim=-1)
+        elif self.reduction == "sum":
+            ret = torch.sum(loss, keepdim=True, dim=-1)
+        else:
+            raise ValueError(self.reduction + " is not valid")
+        return ret
+
+class LpLoss(LossInterface):
+    """
+    The Lp loss implementation class. Creates a criterion that measures
+    the Lp error between each element in the input :math:`x` and
+    target :math:`y`.
+
+    The unreduced (i.e. with :attr:`reduction` set to ``'none'``) loss can
+    be described as:
+
+    .. math::
+        \ell(x, y) = L = \{l_1,\dots,l_N\}^\top, \quad
+        l_n = \left| x_n - y_n \right|^p,
+
+    where :math:`N` is the batch size. If :attr:`reduction` is not ``'none'``
+    (default ``'mean'``), then:
+
+    .. math::
+        \ell(x, y) =
+        \begin{cases}
+            \operatorname{mean}(L), & \text{if reduction} = \text{`mean';}\\
+            \operatorname{sum}(L),  & \text{if reduction} = \text{`sum'.}
+        \end{cases}
+
+    :math:`x` and :math:`y` are tensors of arbitrary shapes with a total
+    of :math:`n` elements each.
+
+    The sum operation still operates over all the elements, and divides by :math:`n`.
+
+    The division by :math:`n` can be avoided if one sets ``reduction = 'sum'``.
+    """
+
+    def __init__(self, p=2, reduction = 'mean'):
+        """
+        :param p: Degree of Lp norm. It specifies the type of norm to
+            be calculated. See :meth:`torch.linalg.norm` ```'ord'``` to
+            see the possible degrees. Default 2 (euclidean norm).
+        :type p: int, optional
+        :param reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction 
+            will be applied, ``'mean'``: the sum of the output will be divided
+            by the number of elements in the output, ``'sum'``: the output will
+            be summed. Note: :attr:`size_average` and :attr:`reduce` are in the
+            process of being deprecated, and in the meantime, specifying either of
+            those two args will override :attr:`reduction`. Default: ``'mean'``.
+        :type reduction: str, optional
+        """
+        super().__init__(reduction=reduction)
+
+        # check consistency
+        check_consistency(p, (str,int,float), 'degree p')
+        self.p = p
+
+    def forward(self, input, target):
+        """Forward method for loss function.
+
+        :param torch.Tensor input: Input tensor from real data.
+        :param torch.Tensor target: Model tensor output.
+        :return: Loss evaluation.
+        :rtype: torch.Tensor
+        """
+        loss = torch.linalg.norm((input-target), ord=self.p, dim=-1)          
```

Comment:
I'm actually extremely lost to understand how to use this loss with physics residual

Reply:
This is just a normal loss. It could happen that we want to use it also without a residual formulation (e.g. for data, or in PID-GAN).

If we want to use it for residual, as we do in `PINN` class, `input=torch.zeros_like(residual)`, `target= residual`